File: Stripe/STPFormTextField.m
//
//  STPFormTextField.m
//  Stripe
//
//  Created by Jack Flintermann on 7/24/15.
//  Copyright (c) 2015 Stripe, Inc. All rights reserved.
//

#import "STPFormTextField.h"

#import "NSString+Stripe.h"
#import "STPBSBNumberValidator.h"
#import "STPCardValidator.h"
#import "STPCardValidator+Private.h"
#import "STPDelegateProxy.h"
#import "STPLocalizationUtils.h"
#import "STPPhoneNumberValidator.h"
#import "STPStringUtils.h"

@interface STPTextFieldDelegateProxy : STPDelegateProxy<UITextFieldDelegate> {
    BOOL _inShouldChangeCharactersInRange;
}
@property (nonatomic, assign) STPFormTextFieldAutoFormattingBehavior autoformattingBehavior;
@property (nonatomic, assign) BOOL selectionEnabled;
@end

@implementation STPTextFieldDelegateProxy

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    if (_inShouldChangeCharactersInRange) {
           // This guards against infinite recursion that happens when moving the cursor
        return YES;
    }
    _inShouldChangeCharactersInRange = YES;

    BOOL insertingIntoEmptyField = (textField.text.length == 0 && range.location == 0 && range.length == 0);
    BOOL hasTextContentType = NO;
    if (@available(iOS 11.0, *)) {
        // This property is available starting in 10.0, but didn't offer in-app suggestions till 11.0
        hasTextContentType = textField.textContentType != nil;
    }

    if (hasTextContentType && insertingIntoEmptyField && [string isEqualToString:@" "]) {
        /* Observed behavior w/iOS 11.0 through 11.2.0 (latest):

         1. UITextContentType suggestions are only available when textField is empty
         2. When user taps a QuickType suggestion for the `textContentType`, UIKit *first*
         calls this method with `range:{0, 0} replacementString:@" "`
         3. If that succeeds (we return YES), this method is called again, this time with
         the actual content to insert (and a space at the end)

         Therefore, always allow entry of a single space in order to support `textContentType`.

         Warning: This bypasses `setText:`, and subsequently `setAttributedText:` and the
         formDelegate methods: `formTextField:modifyIncomingTextChange:` & `formTextFieldTextDidChange:`
         That's acceptable for a single space.
         */
        _inShouldChangeCharactersInRange = NO;
        return YES;
    }

    BOOL deleting = (range.location == textField.text.length - 1 && range.length == 1 && [string isEqualToString:@""]);
    NSString *inputText;
    if (deleting) {
        NSString *sanitized = [self unformattedStringForString:textField.text];
        inputText = [sanitized stp_safeSubstringToIndex:sanitized.length - 1];
    } else {
        NSString *newString = [textField.text stringByReplacingCharactersInRange:range withString:string];
        // Removes any disallowed characters from the whole string.
        // If we (incorrectly) allowed a space to start the text entry hoping it would be a
        // textContentType completion, this will remove it.
        NSString *sanitized = [self unformattedStringForString:newString];
        inputText = sanitized;
    }
    
    UITextPosition *beginning = textField.beginningOfDocument;
    UITextPosition *start = [textField positionFromPosition:beginning offset:range.location];
    
    if ([textField.text isEqualToString:inputText]) {
        _inShouldChangeCharactersInRange = NO;
        return NO;
    }

    textField.text = inputText;

    if (self.autoformattingBehavior == STPFormTextFieldAutoFormattingBehaviorNone && self.selectionEnabled) {
        
        // this will be the new cursor location after insert/paste/typing
        NSInteger cursorOffset = [textField offsetFromPosition:beginning toPosition:start] + string.length;
        
        UITextPosition *newCursorPosition = [textField positionFromPosition:textField.beginningOfDocument offset:cursorOffset];
        UITextRange *newSelectedRange = [textField textRangeFromPosition:newCursorPosition toPosition:newCursorPosition];
        [textField setSelectedTextRange:newSelectedRange];
    }

    _inShouldChangeCharactersInRange = NO;
    return NO;
}

- (NSString *)unformattedStringForString:(NSString *)string {
    switch (self.autoformattingBehavior) {
        case STPFormTextFieldAutoFormattingBehaviorNone:
            return string;
        case STPFormTextFieldAutoFormattingBehaviorCardNumbers:
        case STPFormTextFieldAutoFormattingBehaviorPhoneNumbers:
        case STPFormTextFieldAutoFormattingBehaviorExpiration:
        case STPFormTextFieldAutoFormattingBehaviorBSBNumber:
            return [STPCardValidator sanitizedNumericStringForString:string];
    }
}

@end

typedef NSAttributedString* (^STPFormTextTransformationBlock)(NSAttributedString *inputText);

@interface STPFormTextField()
@property (nonatomic) STPTextFieldDelegateProxy *delegateProxy;
@property (nonatomic, copy) STPFormTextTransformationBlock textFormattingBlock;
@end

@implementation STPFormTextField

+ (NSDictionary *)attributesForAttributedString:(NSAttributedString *)attributedString {
    if (attributedString.length == 0) {
        return @{};
    }
    return [attributedString attributesAtIndex:0 longestEffectiveRange:nil inRange:NSMakeRange(0, attributedString.length)];
}

- (void)setSelectionEnabled:(BOOL)selectionEnabled {
    _selectionEnabled = selectionEnabled;
    self.delegateProxy.selectionEnabled = selectionEnabled;
}

- (void)setAutoFormattingBehavior:(STPFormTextFieldAutoFormattingBehavior)autoFormattingBehavior {
    _autoFormattingBehavior = autoFormattingBehavior;
    self.delegateProxy.autoformattingBehavior = autoFormattingBehavior;
    switch (autoFormattingBehavior) {
        case STPFormTextFieldAutoFormattingBehaviorNone:
        case STPFormTextFieldAutoFormattingBehaviorExpiration:
            self.textFormattingBlock = nil;
            break;
        case STPFormTextFieldAutoFormattingBehaviorCardNumbers:
            self.textFormattingBlock = ^NSAttributedString *(NSAttributedString *inputString) {
                if (![STPCardValidator stringIsNumeric:inputString.string]) {
                    return [inputString copy];
                }
                NSMutableAttributedString *attributedString = [inputString mutableCopy];
                NSArray<NSNumber *> *cardNumberFormat = [STPCardValidator cardNumberFormatForCardNumber:attributedString.string];

                NSUInteger index = 0;
                for (NSNumber *segmentLength in cardNumberFormat) {
                    NSUInteger segmentIndex = 0;
                    for (; index < attributedString.length && segmentIndex < [segmentLength unsignedIntegerValue]; index++, segmentIndex++) {
                        if (index + 1 != attributedString.length && segmentIndex + 1 == [segmentLength unsignedIntegerValue]) {
                            [attributedString addAttribute:NSKernAttributeName value:@(5)
                                                     range:NSMakeRange(index, 1)];
                        } else {
                            [attributedString addAttribute:NSKernAttributeName value:@(0)
                                                     range:NSMakeRange(index, 1)];
                        }
                    }
                }
                return [attributedString copy];
            };
            break;
        case STPFormTextFieldAutoFormattingBehaviorPhoneNumbers: {
            __weak typeof(self) weakSelf = self;
            self.textFormattingBlock = ^NSAttributedString *(NSAttributedString *inputString) {
                if (![STPCardValidator stringIsNumeric:inputString.string]) {
                    return [inputString copy];
                }
                __strong typeof(self) strongSelf = weakSelf;
                NSString *phoneNumber = [STPPhoneNumberValidator formattedSanitizedPhoneNumberForString:inputString.string];
                NSDictionary *attributes = [[strongSelf class] attributesForAttributedString:inputString];
                return [[NSAttributedString alloc] initWithString:phoneNumber attributes:attributes];
            };
        }
            break;
        case STPFormTextFieldAutoFormattingBehaviorBSBNumber: {
            __weak typeof(self) weakSelf = self;
            self.textFormattingBlock = ^NSAttributedString *(NSAttributedString *inputString) {
                if (![STPBSBNumberValidator isStringNumeric:inputString.string]) {
                    return [inputString copy];
                }
                __strong typeof(self) strongSelf = weakSelf;
                NSString *bsbNumber = [STPBSBNumberValidator formattedSantizedTextFromString:inputString.string];
                NSDictionary *attributes = [[strongSelf class] attributesForAttributedString:inputString];
                return [[NSAttributedString alloc] initWithString:bsbNumber attributes:attributes];
            };
        }
            break;
    }
}

- (void)setFormDelegate:(id<STPFormTextFieldDelegate>)formDelegate {
    _formDelegate = formDelegate;
    self.delegate = formDelegate;
}

- (void)insertText:(NSString *)text {
    [self setText:[self.text stringByAppendingString:text]];
}

- (void)setCompressed:(BOOL)compressed {
    if (compressed != _compressed) {
        _compressed = compressed;
        // reset text values as needed
        self.text = self.text;
        self.attributedPlaceholder = self.attributedPlaceholder;
    }
}

- (void)deleteBackward {
    [super deleteBackward];
    if (self.text.length == 0) {
        if ([self.formDelegate respondsToSelector:@selector(formTextFieldDidBackspaceOnEmpty:)]) {
            [self.formDelegate formTextFieldDidBackspaceOnEmpty:self];
        }
    }
}

- (void)setText:(NSString *)text {
    NSString *nonNilText = text ?: @"";
    NSAttributedString *attributed = [[NSAttributedString alloc] initWithString:nonNilText attributes:self.defaultTextAttributes];
    [self setAttributedText:attributed];
}

- (void)setAttributedText:(NSAttributedString *)attributedText {
    NSAttributedString *oldValue = [self attributedText];
    BOOL shouldModify = self.formDelegate && [self.formDelegate respondsToSelector:@selector(formTextField:modifyIncomingTextChange:)];
    NSAttributedString *modified = shouldModify ?
        [self.formDelegate formTextField:self modifyIncomingTextChange:attributedText] :
        attributedText;
    NSAttributedString *transformed = self.textFormattingBlock ? self.textFormattingBlock(modified) : modified;
    [super setAttributedText:transformed];
    [self sendActionsForControlEvents:UIControlEventEditingChanged];
    if ([self.formDelegate respondsToSelector:@selector(formTextFieldTextDidChange:)]) {
        if (![transformed isEqualToAttributedString:oldValue]) {
            [self.formDelegate formTextFieldTextDidChange:self];
        }
    }
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wpartial-availability"
// accessibilityAttributedValue is only defined on iOS 11 and up, but we
// implement it immediately below, so we should just ignore the warning.
- (NSString *)accessibilityValue {
    return [[self accessibilityAttributedValue] string];
}
#pragma clang diagnostic pop

- (NSAttributedString *)accessibilityAttributedValue {
    NSMutableAttributedString *attributedString = [self.attributedText mutableCopy];
    #ifdef __IPHONE_13_0
    if (@available(iOS 13.0, *)) {
        [attributedString addAttribute:UIAccessibilitySpeechAttributeSpellOut value:@(YES) range:NSMakeRange(0, [attributedString length])];
    }
    #endif
    if (!self.validText) {
        NSString *invalidData = STPLocalizedString(@"Invalid data.", @"Spoken during VoiceOver when a form field has failed validation.");
        NSMutableAttributedString *failedString = [[NSMutableAttributedString alloc] initWithString:invalidData attributes:@{UIAccessibilitySpeechAttributePitch: @(0.6)}];
        [failedString appendAttributedString:attributedString];
        attributedString = failedString;
    }
    return attributedString;
}

- (void)setAttributedPlaceholder:(NSAttributedString *)attributedPlaceholder {
    NSAttributedString *transformed = self.textFormattingBlock ? self.textFormattingBlock(attributedPlaceholder) : attributedPlaceholder;
    [super setAttributedPlaceholder:transformed];
}

// Fixes a weird issue related to our custom override of deleteBackwards. This only affects the simulator and iPads with custom keyboards.
- (NSArray *)keyCommands {
    return @[[UIKeyCommand keyCommandWithInput:@"\b" modifierFlags:UIKeyModifierCommand action:@selector(commandDeleteBackwards)]];
}

- (void)commandDeleteBackwards {
    self.text = @"";
}

- (UITextPosition *)closestPositionToPoint:(CGPoint)point {
    if (self.selectionEnabled) {
        return [super closestPositionToPoint:point];
    }
    return [self positionFromPosition:self.beginningOfDocument offset:self.text.length];
}

- (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
    return [super canPerformAction:action withSender:sender] && action == @selector(paste:);
}

- (void)paste:(id)sender {
    if (self.preservesContentsOnPaste) {
        [super paste:sender];
    } else if (self.autoFormattingBehavior == STPFormTextFieldAutoFormattingBehaviorExpiration) {
        self.text = [STPStringUtils expirationDateStringFromString:[UIPasteboard generalPasteboard].string];
    } else {
        self.text = [UIPasteboard generalPasteboard].string;
    }
}

- (void)setDelegate:(id <UITextFieldDelegate>)delegate {
    STPTextFieldDelegateProxy *delegateProxy = [[STPTextFieldDelegateProxy alloc] init];
    delegateProxy.autoformattingBehavior = self.autoFormattingBehavior;
    delegateProxy.selectionEnabled = self.selectionEnabled;
    delegateProxy.delegate = delegate;
    self.delegateProxy = delegateProxy;
    [super setDelegate:delegateProxy];
}

- (id <UITextFieldDelegate>)delegate {
    return self.delegateProxy;
}

@end

File: Stripe/STPFormTextField.swift
//
//  STPFormTextField.swift
//  Stripe
//
//  Created by Jack Flintermann on 7/16/15.
//  Copyright (c) 2015 Stripe, Inc. All rights reserved.
//

import UIKit

enum STPFormTextFieldAutoFormattingBehavior: Int {
    case none
    case phoneNumbers
    case cardNumbers
    case expiration
    case bsbNumber
}

@objc protocol STPFormTextFieldDelegate: UITextFieldDelegate {
    // Note, post-Swift conversion:
    // In lieu of a real delegate proxy, this should always be implemented and call:
    //    if let textField = textField as? STPFormTextField, let delegateProxy = textField.delegateProxy {
    //        return delegateProxy.textField(textField, shouldChangeCharactersIn: range, replacementString: string)
    //    }
    //    return true
    @objc(textField:shouldChangeCharactersInRange:replacementString:) func textField(
        _ textField: UITextField, shouldChangeCharactersIn range: NSRange,
        replacementString string: String
    ) -> Bool

    @objc optional func formTextFieldDidBackspace(onEmpty formTextField: STPFormTextField)
    @objc optional func formTextField(
        _ formTextField: STPFormTextField,
        modifyIncomingTextChange input: NSAttributedString
    ) -> NSAttributedString
    @objc optional func formTextFieldTextDidChange(_ textField: STPFormTextField)
}

@objc class STPFormTextField: STPValidatedTextField {

    private var _selectionEnabled = false
    var selectionEnabled: Bool {
        get {
            _selectionEnabled
        }
        set(selectionEnabled) {
            _selectionEnabled = selectionEnabled
            delegateProxy?.selectionEnabled = selectionEnabled
        }
    }
    /* defaults to NO */
    var preservesContentsOnPaste = false
    // defaults to NO
    private var _compressed = false
    var compressed: Bool {
        get {
            _compressed
        }
        set(compressed) {
            if compressed != _compressed {
                _compressed = compressed
                // reset text values as needed
                _didSetText(text: self.text ?? "")
                _didSetAttributedPlaceholder(attributedPlaceholder: attributedPlaceholder)
            }
        }
    }
    // defaults to NO
    private var _autoFormattingBehavior: STPFormTextFieldAutoFormattingBehavior = .none
    var autoFormattingBehavior: STPFormTextFieldAutoFormattingBehavior {
        get {
            _autoFormattingBehavior
        }
        set(autoFormattingBehavior) {
            _autoFormattingBehavior = autoFormattingBehavior
            delegateProxy?.autoformattingBehavior = autoFormattingBehavior
            switch autoFormattingBehavior {
            case .none, .expiration:
                textFormattingBlock = nil
            case .cardNumbers:
                textFormattingBlock = { inputString in
                    guard let inputString = inputString else {
                        return NSAttributedString()
                    }
                    if !STPCardValidator.stringIsNumeric(inputString.string) {
                        return inputString
                    }
                    let attributedString = NSMutableAttributedString(attributedString: inputString)
                    let cardNumberFormat = STPCardValidator.cardNumberFormat(
                        forCardNumber: attributedString.string)
                    var index = 0
                    for segmentLength in cardNumberFormat {
                        var segmentIndex = 0

                        while index < (attributedString.length)
                            && segmentIndex < Int(segmentLength.uintValue)
                        {
                            if index + 1 != attributedString.length
                                && segmentIndex + 1 == Int(segmentLength.uintValue)
                            {
                                attributedString.addAttribute(
                                    .kern,
                                    value: NSNumber(value: 5),
                                    range: NSRange(location: index, length: 1))
                            } else {
                                attributedString.addAttribute(
                                    .kern,
                                    value: NSNumber(value: 0),
                                    range: NSRange(location: index, length: 1))
                            }

                            index += 1
                            segmentIndex += 1
                        }
                    }
                    return attributedString
                }
            case .phoneNumbers:
                weak var weakSelf = self
                textFormattingBlock = { inputString in
                    if !STPCardValidator.stringIsNumeric(inputString?.string ?? "") {
                        return inputString!
                    }
                    guard let strongSelf = weakSelf else {
                        return inputString!
                    }
                    let phoneNumber = STPPhoneNumberValidator.formattedSanitizedPhoneNumber(
                        for: inputString?.string ?? "")
                    let attributes = type(of: strongSelf).attributes(for: inputString)
                    return NSAttributedString(
                        string: phoneNumber,
                        attributes: attributes as? [NSAttributedString.Key: Any])
                }
            case .bsbNumber:
                weak var weakSelf = self
                textFormattingBlock = { inputString in
                    guard let inputString = inputString else {
                        return NSAttributedString()
                    }
                    if !STPBSBNumberValidator.isStringNumeric(inputString.string) {
                        return inputString
                    }
                    guard let strongSelf = weakSelf else {
                        return NSAttributedString()
                    }
                    let bsbNumber = STPBSBNumberValidator.formattedSanitizedText(
                        from: inputString.string)
                    let attributes = type(of: strongSelf).attributes(for: inputString)
                    return NSAttributedString(
                        string: bsbNumber ?? "",
                        attributes: attributes as? [NSAttributedString.Key: Any])
                }
            }
        }
    }

    private weak var _formDelegate: STPFormTextFieldDelegate?
    weak var formDelegate: STPFormTextFieldDelegate? {
        get {
            _formDelegate
        }
        set(formDelegate) {
            _formDelegate = formDelegate
            delegate = formDelegate
        }
    }
    var delegateProxy: STPTextFieldDelegateProxy?
    private var textFormattingBlock: STPFormTextTransformationBlock?

    class func attributes(for attributedString: NSAttributedString?) -> [AnyHashable: Any]? {
        if attributedString?.length == 0 {
            return [:]
        }
        return attributedString?.attributes(
            at: 0, longestEffectiveRange: nil,
            in: NSRange(location: 0, length: attributedString?.length ?? 0))
    }

    /// :nodoc:
    @objc
    public override func insertText(_ text: String) {
        self.text = self.text ?? "" + text
    }

    /// :nodoc:
    @objc
    public override func deleteBackward() {
        super.deleteBackward()
        if (text?.count ?? 0) == 0 {
            if formDelegate?.responds(
                to: #selector(STPPaymentCardTextField.formTextFieldDidBackspace(onEmpty:))) ?? false
            {
                formDelegate?.formTextFieldDidBackspace?(onEmpty: self)
            }
        }
    }

    /// :nodoc:
    @objc public override var text: String? {
        get {
            return super.text
        }
        set(text) {
            let nonNilText = text ?? ""
            _didSetText(text: nonNilText)
        }
    }
    func _didSetText(text: String) {
        let attributed = NSAttributedString(string: text, attributes: defaultTextAttributes)
        attributedText = attributed
    }

    /// :nodoc:
    @objc public override var attributedText: NSAttributedString? {
        get {
            return super.attributedText
        }
        set(attributedText) {
            let oldValue = self.attributedText
            let shouldModify =
                formDelegate != nil
                && formDelegate?.responds(
                    to: #selector(
                        STPFormTextFieldDelegate.formTextField(_:modifyIncomingTextChange:)))
                    ?? false
            var modified: NSAttributedString?
            if let attributedText = attributedText {
                modified =
                    shouldModify
                    ? formDelegate?.formTextField?(self, modifyIncomingTextChange: attributedText)
                    : attributedText
            }
            let transformed = textFormattingBlock != nil ? textFormattingBlock?(modified) : modified
            super.attributedText = transformed
            sendActions(for: .editingChanged)
            if formDelegate?.responds(
                to: #selector(STPPaymentCardTextField.formTextFieldTextDidChange(_:))) ?? false
            {
                if let oldValue = oldValue {
                    if !(transformed?.isEqual(to: oldValue) ?? false) {
                        formDelegate?.formTextFieldTextDidChange?(self)
                    }
                }
            }
        }
    }

    @objc public override var accessibilityAttributedValue: NSAttributedString? {
        get {
            guard let text = text else {
                return nil
            }
            let attributedString = NSMutableAttributedString(string: text)
            if #available(iOS 13.0, *) {
                attributedString.addAttribute(
                    .accessibilitySpeechSpellOut, value: NSNumber(value: true),
                    range: NSRange(location: 0, length: attributedString.length))
            }
            return attributedString
        }
        set {
            // do nothing
        }
    }

    @objc public override var accessibilityAttributedLabel: NSAttributedString? {
        get {
            guard let accessibilityLabel = accessibilityLabel else {
                return nil
            }
            let attributedString = NSMutableAttributedString(string: accessibilityLabel)
            if !validText {
                let invalidData = STPLocalizedString(
                    "Invalid data.",
                    "Spoken during VoiceOver when a form field has failed validation.")
                let failedString = NSMutableAttributedString(
                    string: invalidData,
                    attributes: [
                        NSAttributedString.Key.accessibilitySpeechPitch: NSNumber(value: 0.6)
                    ])
                attributedString.append(NSAttributedString(string: " "))
                attributedString.append(failedString)
            }
            return attributedString
        }
        set {
            // do nothing
        }
    }

    /// :nodoc:
    @objc public override var attributedPlaceholder: NSAttributedString? {
        get {
            return super.attributedPlaceholder
        }
        set(attributedPlaceholder) {
            _didSetAttributedPlaceholder(attributedPlaceholder: attributedPlaceholder)
        }
    }
    func _didSetAttributedPlaceholder(attributedPlaceholder: NSAttributedString?) {
        let transformed =
            textFormattingBlock != nil
            ? textFormattingBlock?(attributedPlaceholder) : attributedPlaceholder
        super.attributedPlaceholder = transformed
    }

    // Fixes a weird issue related to our custom override of deleteBackwards. This only affects the simulator and iPads with custom keyboards.
    /// :nodoc:
    @objc public override var keyCommands: [UIKeyCommand]? {
        return [
            UIKeyCommand(
                input: "\u{08}", modifierFlags: .command, action: #selector(commandDeleteBackwards))
        ]
    }

    @objc func commandDeleteBackwards() {
        text = ""
    }

    /// :nodoc:
    @objc
    public override func closestPosition(to point: CGPoint) -> UITextPosition? {
        if selectionEnabled {
            return super.closestPosition(to: point)
        }
        return position(from: beginningOfDocument, offset: text?.count ?? 0)
    }

    /// :nodoc:
    @objc
    public override func canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
        return super.canPerformAction(action, withSender: sender) && action == #selector(paste(_:))
    }

    /// :nodoc:
    @objc
    public override func paste(_ sender: Any?) {
        if preservesContentsOnPaste {
            super.paste(sender)
        } else if autoFormattingBehavior == .expiration {
            text = STPStringUtils.expirationDateString(from: UIPasteboard.general.string)
        } else {
            text = UIPasteboard.general.string
        }
    }

    /// :nodoc:
    @objc public override weak var delegate: UITextFieldDelegate? {
        get {
            super.delegate
        }
        set {
            let dProxy = STPTextFieldDelegateProxy()
            dProxy.autoformattingBehavior = autoFormattingBehavior
            dProxy.selectionEnabled = selectionEnabled
            self.delegateProxy = dProxy
            super.delegate = newValue
        }
    }
}

class STPTextFieldDelegateProxy: NSObject, UITextFieldDelegate {
    internal var inShouldChangeCharactersInRange = false
    var autoformattingBehavior: STPFormTextFieldAutoFormattingBehavior = .none
    var selectionEnabled = false

    func textField(
        _ textField: UITextField, shouldChangeCharactersIn range: NSRange,
        replacementString string: String
    ) -> Bool {
        if inShouldChangeCharactersInRange {
            // This guards against infinite recursion that happens when moving the cursor
            return true
        }
        inShouldChangeCharactersInRange = true

        let insertingIntoEmptyField =
            (textField.text?.count ?? 0) == 0 && range.location == 0 && range.length == 0
        let hasTextContentType = textField.textContentType != nil

        if hasTextContentType && insertingIntoEmptyField && (string == " ") {
            /* Observed behavior w/iOS 11.0 through 11.2.0 (latest):

                     1. UITextContentType suggestions are only available when textField is empty
                     2. When user taps a QuickType suggestion for the `textContentType`, UIKit *first*
                     calls this method with `range:{0, 0} replacementString:@" "`
                     3. If that succeeds (we return YES), this method is called again, this time with
                     the actual content to insert (and a space at the end)

                     Therefore, always allow entry of a single space in order to support `textContentType`.

                     Warning: This bypasses `setText:`, and subsequently `setAttributedText:` and the
                     formDelegate methods: `formTextField:modifyIncomingTextChange:` & `formTextFieldTextDidChange:`
                     That's acceptable for a single space.
                     */
            inShouldChangeCharactersInRange = false
            return true
        }

        let deleting =
            range.location == (textField.text?.count ?? 0) - 1 && range.length == 1
            && (string == "")
        var inputText: String?
        if deleting {
            if let sanitized = unformattedString(for: textField.text) {
                inputText = sanitized.stp_safeSubstring(to: sanitized.count - 1)
            }
        } else {
            let newString = (textField.text as NSString?)?.replacingCharacters(
                in: range, with: string)
            // Removes any disallowed characters from the whole string.
            // If we (incorrectly) allowed a space to start the text entry hoping it would be a
            // textContentType completion, this will remove it.
            let sanitized = unformattedString(for: newString)
            inputText = sanitized
        }

        let beginning = textField.beginningOfDocument
        let start = textField.position(from: beginning, offset: range.location)

        if textField.text == inputText {
            inShouldChangeCharactersInRange = false
            return false
        }

        textField.text = inputText

        if autoformattingBehavior == .none && selectionEnabled {

            // this will be the new cursor location after insert/paste/typing
            var cursorOffset: Int?
            if let start = start {
                cursorOffset = textField.offset(from: beginning, to: start) + string.count
            }

            let newCursorPosition = textField.position(
                from: textField.beginningOfDocument, offset: cursorOffset ?? 0)
            var newSelectedRange: UITextRange?
            if let newCursorPosition = newCursorPosition {
                newSelectedRange = textField.textRange(
                    from: newCursorPosition, to: newCursorPosition)
            }
            textField.selectedTextRange = newSelectedRange
        }

        inShouldChangeCharactersInRange = false
        return false
    }

    func unformattedString(for string: String?) -> String? {
        switch autoformattingBehavior {
        case .none:
            return string
        case .cardNumbers, .phoneNumbers, .expiration, .bsbNumber:
            return STPCardValidator.sanitizedNumericString(for: string ?? "")
        }
    }
}

typealias STPFormTextTransformationBlock = (NSAttributedString?) -> NSAttributedString
